File: 0637-average-of-levels-in-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]):
        if not root:
            return []
        storage = [root]
        result = []
        while len(storage) > 0:
            size = len(storage)
            currentVal = 0
            for _ in range(size):
                current = storage.pop(0)
                currentVal += current.val
                if current.left:
                    storage.append(current.left)
                if current.right:
                    storage.append(current.right)
            result.append(currentVal / size)
        return result

File: 0637-average-of-levels-in-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        if (root == nullptr) {
            return ans;
        }
        deque<TreeNode*> storage;
        storage.push_back(root);
    
        while (!storage.empty()) {
            double current_value;
            long long int size = storage.size();
            long long int current_sum = 0;
            for (int i = 0; i < size; i++) {
                TreeNode* CurrentNode = storage.front();
                storage.pop_front();
                current_sum += CurrentNode -> val;
                if (CurrentNode->left != nullptr) {
                    storage.push_back(CurrentNode->left);
                }
                if (CurrentNode->right != nullptr) {
                    storage.push_back(CurrentNode->right);
                }
            }
            ans.push_back(current_sum / (double) size);
            }
        return ans;
    }
};
